buildtools/LibyuiCommon.cmake

```diff
@@ -116,6 +118,64 @@ MACRO( SET_BUILD_FLAGS )	# setup compiler-flags depending on CMAKE_BUILD_TYPE
     )
   ENDIF( NOT BUILD_TYPE_PASSED )
 
+  IF ( ENABLE_TESTS OR ENABLE_CODE_COVERAGE)
+    ENABLE_TESTING()
+    # add a wrapper "tests" target, the builtin "test" cannot be extended :-(
+    ADD_CUSTOM_TARGET(tests
+      make
+      COMMAND make test
+    )
+  ENDIF ( ENABLE_TESTS OR ENABLE_CODE_COVERAGE)
+
+  IF ( ENABLE_CODE_COVERAGE )
+    # pass the coverage options to GCC
+    ADD_DEFINITIONS(--coverage)
+    LINK_LIBRARIES(--coverage)
+    # force debug build to not optimize the code, the optimized out code
+    # might be reported as not covered
+    SET( CMAKE_BUILD_TYPE DEBUG CACHE
+         STRING "set to DEBUG"
+         FORCE
+    )
+
+    FIND_PROGRAM( LCOV_PATH lcov )
+    FIND_PROGRAM( GENHTML_PATH genhtml )
+
+    IF(LCOV_PATH AND GENHTML_PATH)
+      # save the coverage data to coverage/ subdir
+      SET(COVERAGE_DATA "coverage/coverage.info")
+      ADD_CUSTOM_TARGET(coverage
+        # remove the previous coverage data
+        rm -rf coverage/*
+        COMMAND mkdir -p coverage
+        # collect the coverage data, ignore external files (/usr/include/...), ignore tests
+        COMMAND ${LCOV_PATH} --no-external --exclude '*/tests/*' -o ${COVERAGE_DATA} -c -d . -q
+        # generate the HTML coverage report
+        COMMAND ${GENHTML_PATH} -o coverage --legend --title "libyui code coverage" -q ${COVERAGE_DATA}
+        # print the coverage summary on the terminal
+        COMMAND ${LCOV_PATH} --list ${COVERAGE_DATA}
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        COMMENT "Generating the code coverage report..."
+      )
+      # display a comment when finished
+      ADD_CUSTOM_COMMAND(TARGET coverage POST_BUILD
+        COMMAND ;
+        COMMENT "Code coverage gerated to ./coverage/index.html file."
+      )
+      # automatically collect code coverage after "tests" target is finished
+      ADD_CUSTOM_COMMAND(TARGET tests POST_BUILD
+        COMMAND make coverage
+      )
+    ELSE(LCOV_PATH AND GENHTML_PATH)
+      MESSAGE(WARNING "lcov is not installed, code coverage report will not be generated.")
```

Comment:
I'm not sure if this is better than a hard failure.
With `lcov` missing, `make tests` silently skips the coverage test and only with `make coverage` shows me this warning.

Reply:
OK, if it is silently ignored then a hard failure makes sense. I'll update it...